Source/Core/VideoBackends/OGL/VertexManager.cpp

```diff
@@ -109,77 +92,50 @@ void VertexManager::ResetBuffer(u32 stride)
     m_cur_buffer_pointer = m_base_buffer_pointer = buffer.first;
     m_end_buffer_pointer = buffer.first + MAXVBUFFERSIZE;
-    s_baseVertex = buffer.second / stride;
 
-    buffer = s_indexBuffer->Map(MAXIBUFFERSIZE * sizeof(u16));
+    buffer = m_index_buffer->Map(MAXIBUFFERSIZE * sizeof(u16));
     IndexGenerator::Start((u16*)buffer.first);
-    s_index_offset = buffer.second;
   }
 }
 
-void VertexManager::Draw(u32 stride)
+void VertexManager::CommitBuffer(u32 num_vertices, u32 vertex_stride, u32 num_indices,
+                                 u32* out_base_vertex, u32* out_base_index)
 {
-  u32 index_size = IndexGenerator::GetIndexLen();
-  u32 max_index = IndexGenerator::GetNumVerts();
-  GLenum primitive_mode = 0;
+  u32 vertex_data_size = num_vertices * vertex_stride;
+  u32 index_data_size = num_indices * sizeof(u16);
 
-  switch (m_current_primitive_type)
-  {
-  case PrimitiveType::Points:
-    primitive_mode = GL_POINTS;
-    break;
-  case PrimitiveType::Lines:
-    primitive_mode = GL_LINES;
-    break;
-  case PrimitiveType::Triangles:
-    primitive_mode = GL_TRIANGLES;
-    break;
-  case PrimitiveType::TriangleStrip:
-    primitive_mode = GL_TRIANGLE_STRIP;
-    break;
-  }
+  *out_base_vertex = vertex_stride > 0 ? (m_vertex_buffer->GetCurrentOffset() / vertex_stride) : 0;
+  *out_base_index = m_index_buffer->GetCurrentOffset() / sizeof(u16);
 
-  if (g_ogl_config.bSupportsGLBaseVertex)
-  {
-    glDrawRangeElementsBaseVertex(primitive_mode, 0, max_index, index_size, GL_UNSIGNED_SHORT,
-                                  (u8*)nullptr + s_index_offset, (GLint)s_baseVertex);
-  }
-  else
-  {
-    glDrawRangeElements(primitive_mode, 0, max_index, index_size, GL_UNSIGNED_SHORT,
-                        (u8*)nullptr + s_index_offset);
-  }
+  CheckBufferBinding();
+  m_vertex_buffer->Unmap(vertex_data_size);
+  m_index_buffer->Unmap(index_data_size);
 
-  INCSTAT(stats.thisFrame.numDrawCalls);
+  ADDSTAT(stats.thisFrame.bytesVertexStreamed, vertex_data_size);
+  ADDSTAT(stats.thisFrame.bytesIndexStreamed, index_data_size);
 }
 
-void VertexManager::vFlush()
+void VertexManager::UploadConstants()
 {
-  GLVertexFormat* nativeVertexFmt = (GLVertexFormat*)VertexLoaderManager::GetCurrentVertexFormat();
-  u32 stride = nativeVertexFmt->GetVertexStride();
-
-  PrepareDrawBuffers(stride);
-
-  // upload global constants
   ProgramShaderCache::UploadConstants();
+}
 
+void VertexManager::DrawCurrentBatch(u32 base_index, u32 num_indices, u32 base_vertex)
+{
   if (::BoundingBox::active && !g_Config.BBoxUseFragmentShaderImplementation())
   {
     glEnable(GL_STENCIL_TEST);
   }
 
   if (m_current_pipeline_object)
   {
-    g_renderer->SetPipeline(m_current_pipeline_object);
-    Draw(stride);
+    static_cast<Renderer*>(g_renderer.get())->SetPipeline(m_current_pipeline_object);
+    static_cast<Renderer*>(g_renderer.get())->DrawIndexed(base_index, num_indices, base_vertex);
```

Comment:
```
static_cast<Renderer*>(g_renderer.get())->SetPipeline(m_current_pipeline_object);
static_cast<Renderer*>(g_renderer.get())->DrawIndexed(base_index, num_indices, base_vertex);
```
can this simple to
```
g_renderer->DrawIndexed(base_index, num_indices, base_vertex);
```

Reply:
The cast is to avoid the virtual call for DrawIndexed (since it'll always be a GL renderer), but you are correct about the pipeline binding being redundant.